src/main/kotlin/org/rust/ide/annotator/fixes/AddUnsafeFix.kt

```diff
@@ -32,4 +33,16 @@ class AddUnsafeFix(expr: RsBlock) : LocalQuickFixAndIntentionActionOnPsiElement(
             }
         }
     }
+
+    companion object {
+        fun findApplicableBlock(element: PsiElement): RsBlock? {
+            var block = element.ancestorStrict<RsBlock>()
+
+            while (block != null && block.parent !is RsBlockExpr && block.parent !is RsFunction) {
+                block = block.ancestorStrict()
+            }
+
+            return block
+        }
```

Comment:
```suggestion
        fun findApplicableBlock(element: PsiElement): RsBlock? =
            when (val parent = PsiTreeUtil.getParentOfType(element, RsBlockExpr::class.java, RsFunction::class.java)) {
                is RsBlockExpr -> parent.block
                is RsFunction -> parent.block
                else -> null
            }
```

Reply:
And maybe it's better to move this logic into `invoke`?

Reply:
It's in `findApplicableBlock` because `RsDiagnostic.UnsafeError` uses it to find the corresponding block - it is used to set the fix message (`Add unsafe to block` vs `Add unsafe to function`).

Reply:
Ok, then maybe create some factory method instead of `findApplicableContext`? Something like this:

```kotlin
companion object {
    fun create(element: PsiElement): AddUnsafeFix? {
        val parent = PsiTreeUtil.getParentOfType(element, RsBlockExpr::class.java, RsFunction::class.java) ?: return null
        return AddUnsafeFix(parent)
    }
}
```

```kotlin
class AddUnsafeFix private constructor(expr: PsiElement) : LocalQuickFixAndIntentionActionOnPsiElement(expr) {
```

```kotlin
class UnsafeError(
    element: RsExpr,
    private val message: String
) : RsDiagnostic(element) {
    override fun prepare() = PreparedAnnotation(
        ERROR,
        E0133,
        message,
        fixes = listOfNotNull(SurroundWithUnsafeFix(element as RsExpr), AddUnsafeFix.create(element))
    )
}
```

```kotlin
override fun invoke(project: Project, file: PsiFile, editor: Editor?, parent: PsiElement, endElement: PsiElement) {
    val unsafe = RsPsiFactory(project).createUnsafeKeyword()
    when (parent) {
        is RsBlockExpr -> parent.addBefore(unsafe, parent.block)
        is RsFunction -> parent.addBefore(unsafe, parent.fn)
        else -> error("unreachable")
    }
}
```

Reply:
Thanks, that is indeed better.